### PR TITLE
ShareModal: Export options (JSON)

### LINKED ISSUE #85653: Export options (JSON)
**Before**

**JSON**
Screenshot 2024-04-05 at 14 30 08

Screenshot 2024-04-05 at 14 30 24

**Now**

**JSON**
<video src="https://github.com/grafana/grafana/assets/51436795/9ab40337-2a96-44da-be18-a52332d9d494
" />

**We are changing with this task:**

- Having both actions PDF and JSON under an "`Export`" button action - new to our toolbar
- Keeping exporting behavior the same as sharing actions, inside a drawer

**JSON improvements**

- Have the preview JSON at first
- Read-only, can't be edit
- Actions for "Copy to clipboard", and "Download file" with lead icons to be consistent with the PDF feature
- `Toggle` text content changed to **"Export the dashboard to use in another instance"** 

**Resources**
[Figma](https://www.figma.com/file/AjvrAorYpnfSoMpU5kwf39/Adoption-of-Sharing?type=design&node-id=3118-42711&mode=design)

### PR DESCRIPTION
**Pull request by lucychen-grafana**
**What is this feature?**
Created new Export button with Export as JSON option that opens up a drawer for the JSON export feature

**Why do we need this feature?**

Part of the Sharing Modal redesign

**Who is this feature for?**

All users of Grafana Dashboard

**Which issue(s) does this PR fix?**:

Fixes #https://github.com/grafana/grafana/issues/85653

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/features/dashboard-scene/sharing/ExportButton/ExportMenu.tsx [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+    dashboard.showModal(drawer);
+  };
+
+  return (
+    <Menu data-testid={newExportButtonSelector.container}>
+      <Menu.Item
+        testId={newExportButtonSelector.exportAsJson}
+        label="Export as JSON"
+        icon="file-alt"
```
Comment (juanicabanas) [inline public/app/features/dashboard-scene/sharing/ExportButton/ExportMenu.tsx:28]: this should be another icon

**Review by juanicabanas — COMMENTED** [review]
Great job! 👏
I left some comments to work on

**Comment by diegommm** [comment]
Hi! Greetings from the Release Guild. Just letting you know I'm updating your PR's milestone to [11.2.x](https://github.com/grafana/grafana/milestone/529). If you need to see your changes in a previous version, please consider adding a backport label.

Thank you!

File: public/app/features/dashboard-scene/sharing/ExportButton/ExportAsJSON.tsx [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+                language="json"
+                showMiniMap={false}
+                height="500px"
+                width={width}
+                readOnly={true}
+              />
+            );
+          }
+
+          if (dashboardJson.loading) {
```
[the remark this answers is not in the document]
Comment (lucychen-grafana) [reply public/app/features/dashboard-scene/sharing/ExportButton/ExportAsJSON.tsx:116]: Add what?

[the remark this answers is not in the document]
Comment (juanicabanas) [reply public/app/features/dashboard-scene/sharing/ExportButton/ExportAsJSON.tsx:116]: the Spinner component, sorry

File: public/app/features/dashboard-scene/sharing/ExportButton/ExportMenu.tsx (hunk 1 above)
Comment (lucychen-grafana) [reply public/app/features/dashboard-scene/sharing/ExportButton/ExportMenu.tsx:28]: Not sure what you mean by another icon? Is this not the correct one?

File: public/app/features/dashboard-scene/sharing/ExportButton/ExportAsJson.tsx [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+const selector = e2eSelectors.pages.ExportDashboardDrawer.ExportAsJson;
+
+export class ExportAsJson extends ShareExportTab {
+  static Component = ExportAsJsonRenderer;
+}
+
+function ExportAsJsonRenderer({ model }: SceneComponentProps<ExportAsJson>) {
+  const theme = useTheme2();
+  const styles = getStyles(theme);
```
Comment (juanicabanas) [inline public/app/features/dashboard-scene/sharing/ExportButton/ExportAsJson.tsx:22]:
```suggestion
const styles = useStyles2(getStyles);
```
this already adds the theme inside

File: public/app/features/dashboard-scene/sharing/ExportButton/ExportAsJson.tsx [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  }, [isSharingExternally]);
+
+  return (
+    <>
+      <p>
+        <Trans i18nKey="export.json.info-text">
+          Copy or download a JSON file containing the JSON of your dashboard.
+        </Trans>
+      </p>
+
```
Comment (juanicabanas) [inline public/app/features/dashboard-scene/sharing/ExportButton/ExportAsJson.tsx:42]: remove blank line

File: public/app/features/dashboard-scene/sharing/ExportButton/ExportAsJson.tsx [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+          data-testid={selector.exportExternallyToggle}
+          id="export-externally-toggle"
+          value={isSharingExternally}
+          onChange={model.onShareExternallyChange}
+        />
+        <Label>
+          <Trans i18nKey="export.json.export-externally-label">Export the dashboard to use in another instance</Trans>
+        </Label>
+      </Stack>
+
```
Comment (juanicabanas) [inline public/app/features/dashboard-scene/sharing/ExportButton/ExportAsJson.tsx:54]: same here

File: public/app/features/dashboard-scene/sharing/ExportButton/ExportAsJson.tsx [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+                width={width}
+                readOnly={true}
+              />
+            );
+          }
+
+          return dashboardJson.loading && <Spinner />;
+        }}
+      </AutoSizer>
+
```
Comment (juanicabanas) [inline public/app/features/dashboard-scene/sharing/ExportButton/ExportAsJson.tsx:73]: same here

File: public/app/features/dashboard-scene/scene/NavToolbarActions.tsx [hunk 7]
```diff
@@ -578,6 +579,12 @@ export function ToolbarActions({ dashboard }: Props) {
     },
   });
 
+  toolbarActions.push({
+    group: 'new-export-dashboard-button',
+    condition: config.featureToggles.newDashboardSharingComponent && showShareButton,
+    render: () => <ExportButton key="new-export-dashboard-button" dashboard={dashboard} />,
+  });
+
```
Comment (juanicabanas) [inline public/app/features/dashboard-scene/scene/NavToolbarActions.tsx:587]: the export button is before Share button and is part of the same group so it doesn't have boundaries between each other (I think that's the group prop, so we should change to be the same)
image

File: public/app/features/dashboard-scene/sharing/ExportButton/ExportAsJson.tsx [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+      <Stack gap={1} flex={1} direction={{ xs: 'column', sm: 'row' }}>
```
Comment (juanicabanas) [inline public/app/features/dashboard-scene/sharing/ExportButton/ExportAsJson.tsx:74]: great! Take a look at ShareExternally that it also adds a paddingBottom

File: public/app/features/dashboard-scene/sharing/ExportButton/ExportButton.tsx [hunk 9]
```diff
@@ -0,0 +1,39 @@
+import React, { useCallback, useState } from 'react';
+// import { useAsyncFn } from 'react-use';
```
Comment (juanicabanas) [inline public/app/features/dashboard-scene/sharing/ExportButton/ExportButton.tsx:2]: remove this

File: public/app/features/dashboard-scene/sharing/ExportButton/ExportButton.tsx [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+  return (
+    <ButtonGroup data-testid={newExportButtonSelector.container}>
+      <Dropdown overlay={MenuActions} placement="bottom-end" onVisibleChange={onMenuClick}>
+        <Button
+          data-testid={newExportButtonSelector.arrowMenu}
+          size="sm"
+          variant="secondary"
+          fill="solid"
+          tooltip="Export"
```
Comment (juanicabanas) [inline public/app/features/dashboard-scene/sharing/ExportButton/ExportButton.tsx:31]: translate this

File: public/app/features/dashboard-scene/sharing/ExportButton/ExportButton.tsx [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+        >
+          <Trans i18nKey="export.menu.export-as-json-label">Export</Trans>
+          <Icon name={isOpen ? 'angle-up' : 'angle-down'} size="lg" />
```
Comment (juanicabanas) [inline public/app/features/dashboard-scene/sharing/ExportButton/ExportButton.tsx:34]: I think there should be more space between Export and the icon?

File: public/app/features/dashboard-scene/sharing/ShareExportTab.tsx [hunk 12]
```diff
@@ -70,9 +70,9 @@ export class ShareExportTab extends SceneObjectBase<ShareExportTabState> {
       : saveModel;
 
     return exportable;
-  }
+  };
 
-  public async onSaveAsFile() {
+  public onSaveAsFile = async () => {
```
Comment (juanicabanas) [inline public/app/features/dashboard-scene/sharing/ShareExportTab.tsx:75]: great!

**Review by juanicabanas — COMMENTED** [review]
Left some small comments and other nits

File: public/app/features/dashboard-scene/sharing/ExportButton/ExportAsJson.tsx (hunk 8 above)
Comment (lucychen-grafana) [reply public/app/features/dashboard-scene/sharing/ExportButton/ExportAsJson.tsx:74]: why do we need paddingBottom? There is nothing after the buttons

Comment (juanicabanas) [reply public/app/features/dashboard-scene/sharing/ExportButton/ExportAsJson.tsx:74]: it adds space after scrolling in responsive view

File: public/app/features/dashboard-scene/scene/NavToolbarActions.test.tsx [hunk 13]
```diff
@@ -162,6 +162,12 @@ describe('NavToolbarActions', () => {
       const newShareButton = screen.getByTestId(selectors.pages.Dashboard.DashNav.newShareButton.container);
       expect(newShareButton).toBeInTheDocument();
     });
+    it('Should show new export button when newDashboardSharingComponent FF is enabled', async () => {
+      config.featureToggles.newDashboardSharingComponent = true;
+      setup();
+      const newExportButton = screen.getByTestId(selectors.pages.Dashboard.DashNav.NewExportButton.container);
+      expect(newExportButton).toBeInTheDocument();
```
Comment (juanicabanas) [inline public/app/features/dashboard-scene/scene/NavToolbarActions.test.tsx:169]: great! Could you add in the first test, above, to not expect this button when it's the old design? 
In this test
`Should show old share button when newDashboardSharingComponent FF is disabled`

File: public/app/features/dashboard-scene/sharing/ExportButton/ExportAsJson.tsx [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+
+  return (
+    <>
+      <p>
+        <Trans i18nKey="export.json.info-text">
+          Copy or download a JSON file containing the JSON of your dashboard
+        </Trans>
+      </p>
+      <Stack gap={1} alignItems="center">
+        <Switch
```
Comment (juanicabanas) [inline public/app/features/dashboard-scene/sharing/ExportButton/ExportAsJson.tsx:39]: also add the label here as prop

**Review by juanicabanas — APPROVED** [review]
Great work!
I would like to add that line in the test I mentioned and then, a little nit.
LGTM. Approved 🚀

**Pull request merged**